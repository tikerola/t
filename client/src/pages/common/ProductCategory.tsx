import React, { Dispatch, SetStateAction } from "react";
import { Item } from "../../hooks/useRequestData";

import { faSearch } from "@fortawesome/free-solid-svg-icons";
import {
  Button,
  ButtonContainer,
  ButtonContainerBackground,
  PageNumberText,
  Title,
  Input,
  InputContainer,
  Icon,
  NoMatchesText,
  NoMatchesTextContainer,
} from "./ProductCategory.styles";
import { RenderProductData } from "./RenderProductData";
import { PRODUCTS_PER_PAGE } from "./constants";
import { BurgerMenuContext } from "../../App";

interface IProps {
  title: string;
  page: number;
  setPage: Dispatch<SetStateAction<number>>;
  productData: Item[];
  numOfProducts: number;
  setFilter: Dispatch<SetStateAction<string>>;
}

export const ProductCategory = ({
  title,
  page,
  setPage,
  productData,
  numOfProducts,
  setFilter,
}: IProps): JSX.Element => {
  const { open } = React.useContext(BurgerMenuContext);

  const handlePageChange = (direction: string): void => {
    if (direction === "previous" && page > 1)
      setPage((prevPage) => prevPage - 1);
    else if (direction === "next" && getPaginationEnd() < numOfProducts)
      setPage((prevPage) => prevPage + 1);
  };

  const getPaginationStart = (): number => {
    return (page - 1) * PRODUCTS_PER_PAGE + 1;
  };

  const getPaginationEnd = (): number => {
    const end = (page - 1) * PRODUCTS_PER_PAGE + PRODUCTS_PER_PAGE;
    return end < numOfProducts ? end : numOfProducts;
  };

  return (
    <div>
      <Title>{title}</Title>
      <InputContainer open={open}>
        <Icon icon={faSearch} color="#bbb" />
        <Input
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
            setPage(1);
            setFilter(e.target.value);
          }}
          placeholder="Filter by Name"
        />
      </InputContainer>

      <RenderProductData productData={productData} />
      {productData.length > 0 ? (
        <ButtonContainerBackground>
          <ButtonContainer>
            <Button onClick={() => handlePageChange("previous")}>
              {"<< "}Prev
            </Button>
            <PageNumberText>
              {getPaginationStart()} - {getPaginationEnd()} / {numOfProducts}
            </PageNumberText>
            <Button onClick={() => handlePageChange("next")}>
              Next{" >>"}
            </Button>
          </ButtonContainer>
        </ButtonContainerBackground>
      ) : (
        <NoMatchesTextContainer>
          <NoMatchesText>No Matches Found!</NoMatchesText>
        </NoMatchesTextContainer>
      )}
    </div>
  );
};
