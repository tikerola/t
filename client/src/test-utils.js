import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { lightTheme } from "./styles/theme";
import { Router } from "react-router-dom";

import { createMemoryHistory } from "history";

const history = createMemoryHistory();
const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider theme={lightTheme}>
      <Router history={history}>{children}</Router>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
