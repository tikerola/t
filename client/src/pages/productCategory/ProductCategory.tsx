import React from "react";
import { useRequestData } from "../../hooks/useRequestData";
import { renderProductData } from "../commonJSX/renderProductData";
import { Title } from "./ProductCategory.styles";

type Props = {
  title: string;
  category: string;
};

export const ProductCategory = ({ title, category }: Props): JSX.Element => {
  const [productData] = useRequestData({ url: `/products/${category}` });

  console.log(productData, "piip");

  return (
    <div>
      <Title>{title}</Title>
      {renderProductData(productData)}
    </div>
  );
};
