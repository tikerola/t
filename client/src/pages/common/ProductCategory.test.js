import React from "react";
import { render, fireEvent, waitFor } from "../../test-utils";
import { ProductCategory } from "./ProductCategory";
import userEvent from "@testing-library/user-event";
import { PRODUCTS_PER_PAGE } from "./constants";

jest.mock("./constants", () => ({
  PRODUCTS_PER_PAGE: 2,
}));

const productData = [
  {
    id: "f33561de3a864f951a",
    type: "jackets",
    name: "EWHHOP ROOM",
    color: ["blue"],
    price: 52,
    manufacturer: "reps",
  },
  {
    id: "0e4772c827c4296592fbd",
    type: "jackets",
    name: "WEERLEP METROPOLIS RAPTOR",
    color: ["black"],
    price: 98,
    manufacturer: "reps",
  },
  {
    id: "6d39a08b3bcae88a67",
    type: "jackets",
    name: "DERWEER TYRANNUS BANG",
    color: ["purple"],
    price: 15,
    manufacturer: "abiplos",
  },
];

describe("<ProductCategory/>", () => {
  it("should match snapshot", async () => {
    const setPageMock = jest.fn();
    const setFilterMock = jest.fn();
    const component = render(
      <ProductCategory
        title="Jackets"
        page={1}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={productData}
      />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should test filter input", async () => {
    const setPageMock = jest.fn();
    const setFilterMock = jest.fn();
    const component = render(
      <ProductCategory
        title="Jackets"
        page={1}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={productData}
      />
    );

    const input = component.queryByPlaceholderText(/Filter by Name/i);
    userEvent.type(input, "hello");

    expect(setFilterMock).toHaveBeenCalledWith("hello");
    expect(setFilterMock).toHaveBeenCalledTimes(5);
  });

  it("should render pagination when productData exists", async () => {
    const setPageMock = jest.fn();
    const setFilterMock = jest.fn();
    const component = render(
      <ProductCategory
        title="Jackets"
        page={1}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={productData}
      />
    );

    const next = component.queryByText("Next >>");
    expect(next).toBeDefined();
  });

  it("should not render pagination when productData is empty", async () => {
    const setPageMock = jest.fn();
    const setFilterMock = jest.fn();
    const component = render(
      <ProductCategory
        title="Jackets"
        page={1}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={[]}
      />
    );

    const next = component.queryByText("Next >>");
    expect(next).not.toBeTruthy();

    const noMatchesFound = component.queryByText("No Matches Found!");
    expect(noMatchesFound).toBeDefined();
  });

  it("should be able to change next and previous page only when appropriate", async () => {
    const setPageMock = jest.fn();
    const setFilterMock = jest.fn();
    const component = render(
      <ProductCategory
        title="Jackets"
        page={1}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={productData}
      />
    );

    const next = component.queryByText("Next >>");
    const previous = component.queryByText("<< Prev");

    let paginationText = component.queryByText("1 - 2 / 3");
    expect(paginationText).toBeDefined();

    fireEvent.click(previous);
    paginationText = component.queryByText("1 - 2 / 3");
    expect(paginationText).toBeDefined();

    expect(setPageMock).not.toHaveBeenCalled();

    fireEvent.click(next);
    paginationText = component.queryByText("3 - 3 / 3");
    expect(paginationText).toBeDefined();

    component.rerender(
      <ProductCategory
        title="Jackets"
        page={2}
        setPage={setPageMock}
        setFilter={setFilterMock}
        numOfProducts={3}
        productData={productData}
      />
    );

    fireEvent.click(next);
    paginationText = component.queryByText("3 - 3 / 3");
    expect(paginationText).toBeDefined();

    fireEvent.click(previous);

    paginationText = component.queryByText("1 - 2 / 3");
    expect(paginationText).toBeDefined();

    expect(setPageMock).toHaveBeenCalledTimes(2);
  });
});
