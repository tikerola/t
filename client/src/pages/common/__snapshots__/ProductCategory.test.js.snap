// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCategory/> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="sc-bdfBwQ fLjPHF"
    >
      Jackets
    </h1>
    <div
      class="sc-gsTCUz rJKlv"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-search fa-w-16 sc-dlfnbm hUuCRR"
        color="#bbb"
        data-icon="search"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          fill="currentColor"
        />
      </svg>
      <input
        class="sc-hKgILt cTcqRn"
        placeholder="Filter by Name"
      />
    </div>
    <div
      class="sc-eCssSg duHnSf"
    >
      <div
        class="sc-hBEYos mHOcC"
      >
        <div
          class="sc-jSgupP dyMIlS"
        >
          EWHHOP ROOM
        </div>
        <div
          class="sc-gKsewC iaRKPh"
        >
          reps
        </div>
        <div
          class="sc-pFZIQ kzHLGl"
        >
          Available colors: 
          <div
            class="sc-jrAGrp iyIsys"
            color="blue"
            data-testid="color-button"
          />
          <div
            class="sc-iBPRYJ iXdAIx"
          />
        </div>
        <div
          class="sc-fubCfw bDUSrg"
        >
          $52
        </div>
      </div>
      <div
        class="sc-hBEYos mHOcC"
      >
        <div
          class="sc-jSgupP dyMIlS"
        >
          WEERLEP METROPOLIS RAPTOR
        </div>
        <div
          class="sc-gKsewC iaRKPh"
        >
          reps
        </div>
        <div
          class="sc-pFZIQ kzHLGl"
        >
          Available colors: 
          <div
            class="sc-jrAGrp fRONuR"
            color="black"
            data-testid="color-button"
          />
          <div
            class="sc-iBPRYJ iXdAIx"
          />
        </div>
        <div
          class="sc-fubCfw fnFdVi"
        >
          $98
        </div>
      </div>
      <div
        class="sc-hBEYos mHOcC"
      >
        <div
          class="sc-jSgupP dyMIlS"
        >
          DERWEER TYRANNUS BANG
        </div>
        <div
          class="sc-gKsewC iaRKPh"
        >
          abiplos
        </div>
        <div
          class="sc-pFZIQ kzHLGl"
        >
          Available colors: 
          <div
            class="sc-jrAGrp efDoro"
            color="purple"
            data-testid="color-button"
          />
          <div
            class="sc-iBPRYJ iXdAIx"
          />
        </div>
        <div
          class="sc-fubCfw iQukNH"
        >
          $15
        </div>
      </div>
    </div>
    <div
      class="sc-kEjbxe jJfbkl"
    >
      <div
        class="sc-iqHYGH kPMMMG"
      >
        <div
          class="sc-crrsfI jdXTAL"
        >
          &lt;&lt; Prev
        </div>
        <div
          class="sc-dQppl iGUJzW"
        >
          1 - 2 / 3
        </div>
        <div
          class="sc-crrsfI jdXTAL"
        >
          Next &gt;&gt;
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
